File: backend/app/services/radar_simulator.py
"""Radar simulation service - generates realistic radar targets continuously."""

import asyncio
import numpy as np
from datetime import datetime
from typing import List, Dict, Any
from app.models.schemas import RadarScan, RadarTarget
from app.core.logging import radar_logger
import uuid
import time


class RadarSimulator:
    """Simulates continuous radar scanning with realistic target motion."""
    
    TARGET_TYPES = ["DRONE", "AIRCRAFT", "HELICOPTER", "BIRD", "MISSILE", "UNKNOWN"]
    
    def __init__(self):
        self.scan_count = 0
        self.last_scan_time = None
        self.active_targets: Dict[str, Dict[str, Any]] = {}
        self.simulation_mode = True
        self.simulation_speed = 1.0  # Speed multiplier for target motion
        self.target_counter = 0  # For generating meaningful target names
    
    async def _generate_new_targets(self) -> int:
        """Generate some new targets randomly."""
        num_new = np.random.randint(1, 4)
        
        for _ in range(num_new):
            self.target_counter += 1
            target_type = np.random.choice(self.TARGET_TYPES)
            target_name = f"{target_type}-{self.target_counter}"
            target_id = target_name  # Use human-readable name as ID
            
            self.active_targets[target_id] = {
                "name": target_name,
                "type": target_type,
                "x": np.random.uniform(-500, 500),
                "y": np.random.uniform(-500, 500),
                "vx": np.random.uniform(-30, 30),
                "vy": np.random.uniform(-30, 30),
                "strength": np.random.uniform(0.4, 1.0),
                "created_at": time.time(),
                "hits": 1
            }
        
        return num_new
    
    async def _update_target_positions(self):
        """Move active targets slightly (simulate tracking)."""
        for target_id, target in list(self.active_targets.items()):
            # Update position
            dt = 0.3  # Timestep in seconds
            target["x"] += target["vx"] * dt * self.simulation_speed
            target["y"] += target["vy"] * dt * self.simulation_speed
            
            # Bounce off boundaries
            if target["x"] > 500 or target["x"] < -500:
                target["vx"] *= -1
            if target["y"] > 500 or target["y"] < -500:
                target["vy"] *= -1
            
            # Clamp position
            target["x"] = np.clip(target["x"], -500, 500)
            target["y"] = np.clip(target["y"], -500, 500)
            
            # Randomly remove old targets (5% chance per cycle)
            if np.random.random() < 0.05:
                del self.active_targets[target_id]
            else:
                target["hits"] += 1
    
    async def scan(self) -> RadarScan:
        """Execute simulated radar scan."""
        self.scan_count += 1
        scan_id = str(uuid.uuid4())
        now = datetime.utcnow()
        self.last_scan_time = now
        
        # Update positions of active targets
        await self._update_target_positions()
        
        # Generate some new targets
        num_new = await self._generate_new_targets()
        
        # Ensure minimum targets
        total_targets = max(1, len(self.active_targets))
        
        radar_data = RadarScan(
            scan_id=scan_id,
            timestamp=now,
            frame_count=512,
            targets_detected=total_targets,
            signal_strength=np.random.uniform(0.7, 1.0),
            noise_level=np.random.uniform(0.05, 0.15)
        )
        
        radar_logger.log_event(
            "scan_complete",
            "radar_simulator",
            {
                "scan_id": scan_id,
                "targets": total_targets,
                "new_targets": num_new,
                "active_count": len(self.active_targets)
            },
            level="INFO"
        )
        
        return radar_data
    
    async def get_targets_from_scan(self, scan_id: str) -> List[RadarTarget]:
        """Extract simulated targets from scan."""
        targets = []
        
        for target_id, target_data in self.active_targets.items():
            # Calculate range and bearing from position
            range_m = np.sqrt(target_data["x"]**2 + target_data["y"]**2) * 100 + np.random.uniform(100, 1000)
            bearing_deg = np.degrees(np.arctan2(target_data["y"], target_data["x"])) % 360
            velocity_mps = np.sqrt(target_data["vx"]**2 + target_data["vy"]**2) * 10
            
            target = RadarTarget(
                id=target_id,
                range_m=range_m,
                bearing_deg=bearing_deg,
                velocity_mps=velocity_mps,
                rcs_dbsm=np.random.uniform(-15, 10),
                signal_strength=np.clip(target_data["strength"], 0.4, 1.0),
                confidence=0.8 + (target_data["hits"] * 0.02),
                timestamp=datetime.utcnow()
            )
            targets.append(target)
        
        # Ensure at least 1 target
        if not targets:
            self.target_counter += 1
            target_type = np.random.choice(self.TARGET_TYPES)
            target_name = f"{target_type}-{self.target_counter}"
            target_id = target_name
            
            self.active_targets[target_id] = {
                "name": target_name,
                "type": target_type,
                "x": np.random.uniform(-200, 200),
                "y": np.random.uniform(-200, 200),
                "vx": np.random.uniform(-20, 20),
                "vy": np.random.uniform(-20, 20),
                "strength": np.random.uniform(0.6, 0.95),
                "created_at": time.time(),
                "hits": 1
            }
            
            target = RadarTarget(
                id=target_id,
                range_m=np.random.uniform(1000, 50000),
                bearing_deg=np.random.uniform(0, 360),
                velocity_mps=np.random.uniform(0, 100),
                rcs_dbsm=np.random.uniform(-10, 15),
                signal_strength=0.85,
                confidence=0.85,
                timestamp=datetime.utcnow()
            )
            targets.append(target)
        
        radar_logger.log_event(
            "targets_extracted",
            "radar_simulator",
            {"scan_id": scan_id, "target_count": len(targets)},
            level="INFO"
        )
        
        return targets
    
    async def get_signal_quality(self) -> Dict[str, float]:
        """Get simulated signal quality metrics."""
        return {
            "snr_db": np.random.uniform(15, 35),
            "noise_floor_dbm": np.random.uniform(-95, -85),
            "peak_signal_dbm": np.random.uniform(-30, 10),
            "system_temperature_k": 290.0 + np.random.uniform(-5, 5)
        }
    
    def get_active_target_count(self) -> int:
        """Get number of active simulated targets."""
        return len(self.active_targets)

File: frontend/src/utils/radarSimulator.ts
/**
 * Frontend Radar Simulation - Fallback when backend is unavailable
 * Generates realistic moving targets to ensure UI never goes blank
 */

export interface SimulatedTarget {
  id: string
  name: string
  type: 'DRONE' | 'AIRCRAFT' | 'HELICOPTER' | 'BIRD' | 'MISSILE' | 'UNKNOWN'
  x: number
  y: number
  velocity: number
  heading: number
  strength: number
  threat_level: 'LOW' | 'MEDIUM' | 'HIGH'
  timestamp: string
}

export interface SimulatedMetrics {
  cycle: number
  timestamp: string
  targets_detected: number
  threats_high: number
  threats_medium: number
  threats_low: number
  cpu_usage: number
  memory_usage: number
  cycle_time_ms: number
}

class SimulatedTarget_ implements SimulatedTarget {
  id: string
  name: string
  type: 'DRONE' | 'AIRCRAFT' | 'HELICOPTER' | 'BIRD' | 'MISSILE' | 'UNKNOWN'
  x: number
  y: number
  vx: number
  vy: number
  velocity: number
  heading: number
  strength: number
  threat_level: 'LOW' | 'MEDIUM' | 'HIGH' = 'LOW'
  created_at: number
  timestamp: string

  private static targetCounter = 0
  private static typeNames = ['DRONE', 'AIRCRAFT', 'HELICOPTER', 'BIRD', 'MISSILE'] as const

  constructor() {
    const types = SimulatedTarget_.typeNames
    this.type = types[Math.floor(Math.random() * types.length)]
    SimulatedTarget_.targetCounter++
    this.id = `sim_${SimulatedTarget_.targetCounter}`
    this.name = `${this.type}-${SimulatedTarget_.targetCounter}`
    this.x = Math.random() * 1000 - 500
    this.y = Math.random() * 1000 - 500
    this.vx = Math.random() * 60 - 30
    this.vy = Math.random() * 60 - 30
    this.velocity = Math.sqrt(this.vx ** 2 + this.vy ** 2)
    this.heading = (Math.atan2(this.y, this.x) * 180) / Math.PI
    this.strength = Math.random() * 0.5 + 0.5
    this.created_at = Date.now()
    this.timestamp = new Date().toISOString()
  }

  update(dt: number = 0.3) {
    // Add slight acceleration
    this.vx += (Math.random() - 0.5) * 10
    this.vy += (Math.random() - 0.5) * 10

    // Clamp velocity
    const speed = Math.sqrt(this.vx ** 2 + this.vy ** 2)
    if (speed > 40) {
      this.vx = (this.vx / speed) * 40
      this.vy = (this.vy / speed) * 40
    }

    // Update position
    this.x += this.vx * dt
    this.y += this.vy * dt

    // Boundary wrap
    if (this.x > 500 || this.x < -500) this.vx *= -1
    if (this.y > 500 || this.y < -500) this.vy *= -1

    this.x = Math.max(-500, Math.min(500, this.x))
    this.y = Math.max(-500, Math.min(500, this.y))

    // Update derived values
    this.velocity = Math.sqrt(this.vx ** 2 + this.vy ** 2)
    this.heading = (Math.atan2(this.y, this.x) * 180) / Math.PI
    this.strength = Math.max(0.3, Math.min(1.0, this.strength + (Math.random() - 0.5) * 0.1))

    // Randomly set threat level
    if (Math.random() < 0.1) {
      const r = Math.random()
      if (r < 0.7) this.threat_level = 'LOW'
      else if (r < 0.9) this.threat_level = 'MEDIUM'
      else this.threat_level = 'HIGH'
    }

    this.timestamp = new Date().toISOString()
  }

  toJSON(): SimulatedTarget {
    return {
      id: this.id,
      name: this.name,
      type: this.type,
      x: this.x,
      y: this.y,
      velocity: this.velocity,
      heading: this.heading,
      strength: this.strength,
      threat_level: this.threat_level,
      timestamp: this.timestamp
    }
  }
}

export class RadarSimulator {
  private targets: SimulatedTarget_[] = []
  private cycleCount = 0
  private metricsHistory: SimulatedMetrics[] = []
  private lifespan = 50

  executeSimulationCycle(): {
    targets: SimulatedTarget[]
    metrics: SimulatedMetrics
  } {
    this.cycleCount++

    // Remove old targets
    this.targets = this.targets.filter(
      (t) => (Date.now() - t.created_at) < this.lifespan * 300
    )

    // Update existing targets
    this.targets.forEach((t) => t.update())

    // Spawn 2-6 new targets
    const numNew = Math.floor(Math.random() * 5) + 2
    for (let i = 0; i < numNew && this.targets.length < 50; i++) {
      this.targets.push(new SimulatedTarget_())
    }

    // Generate metrics
    const metrics: SimulatedMetrics = {
      cycle: this.cycleCount,
      timestamp: new Date().toISOString(),
      targets_detected: this.targets.length,
      threats_high: this.targets.filter((t) => t.threat_level === 'HIGH').length,
      threats_medium: this.targets.filter((t) => t.threat_level === 'MEDIUM').length,
      threats_low: this.targets.filter((t) => t.threat_level === 'LOW').length,
      cpu_usage: 15 + (Math.random() - 0.5) * 10,
      memory_usage: 250 + (Math.random() - 0.5) * 40,
      cycle_time_ms: 2.5 + (Math.random() - 0.5) * 2
    }

    this.metricsHistory.push(metrics)
    if (this.metricsHistory.length > 100) {
      this.metricsHistory = this.metricsHistory.slice(-100)
    }

    return {
      targets: this.targets.map((t) => t.toJSON()),
      metrics
    }
  }

  getMetricsHistory(): SimulatedMetrics[] {
    return this.metricsHistory.slice(-50)
  }
}
